stratosphere/pm/source/pm_main.cpp

```diff
@@ -69,30 +69,35 @@ void __libnx_initheap(void) {
     fake_heap_end   = (char*)addr + size;
 }
 
-void RegisterPrivilegedProcessesWithFs() {
-    /* Ensures that all privileged processes are registered with full FS permissions. */
-    constexpr u64 PRIVILEGED_PROCESS_MIN = 0;
-    constexpr u64 PRIVILEGED_PROCESS_MAX = 0x4F;
-
-    const u32 PRIVILEGED_FAH[0x1C/sizeof(u32)] = {0x00000001, 0x00000000, 0x80000000, 0x0000001C, 0x00000000, 0x0000001C, 0x00000000};
-    const u32 PRIVILEGED_FAC[0x2C/sizeof(u32)] = {0x00000001, 0x00000000, 0x80000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF};
-
-    u32 num_pids;
-    u64 pids[PRIVILEGED_PROCESS_MAX+1];
-    if (R_SUCCEEDED(svcGetProcessList(&num_pids, pids, sizeof(pids)/sizeof(pids[0])))) {
-        for (u32 i = 0; i < num_pids; i++) {
-            const u64 pid = pids[i];
-            if (PRIVILEGED_PROCESS_MIN <= pid && pid <= PRIVILEGED_PROCESS_MAX) {
-                fsprUnregisterProgram(pid);
-                fsprRegisterProgram(pid, pid, FsStorageId_NandSystem,  PRIVILEGED_FAH, sizeof(PRIVILEGED_FAH), PRIVILEGED_FAC, sizeof(PRIVILEGED_FAC));
+namespace {
+
+    static constexpr u32 PrivilegedFileAccessHeader[0x1C / sizeof(u32)]  = {0x00000001, 0x00000000, 0x80000000, 0x0000001C, 0x00000000, 0x0000001C, 0x00000000};
+    static constexpr u32 PrivilegedFileAccessControl[0x2C / sizeof(u32)] = {0x00000001, 0x00000000, 0x80000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF};
+    static constexpr size_t ProcessCountMax = 0x40;
+
+    /* This works around a bug fixed by FS in 4.0.0. */
+    /* Not doing so will cause KIPs with higher process IDs than 7 to be unable to use filesystem services. */
+    void RegisterPrivilegedProcessWithFs(u64 process_id) {
+        fsprUnregisterProgram(process_id);
+        fsprRegisterProgram(process_id, process_id, FsStorageId_NandSystem, PrivilegedFileAccessHeader, sizeof(PrivilegedFileAccessHeader), PrivilegedFileAccessControl, sizeof(PrivilegedFileAccessControl));
+    }
+
+    void RegisterPrivilegedProcessesWithFs() {
+        /* Get privileged process range. */
+        u64 min_priv_process_id = 0, max_priv_process_id = 0;
+        sts::cfg::GetInitialProcessRange(&min_priv_process_id, &max_priv_process_id);
+
+        /* Get list of processes, register all privileged ones. */
+        u32 num_pids;
+        u64 pids[ProcessCountMax];
+        R_ASSERT(svcGetProcessList(&num_pids, pids, ProcessCountMax));
+        for (size_t i = 0; i < num_pids; i++) {
+            if (min_priv_process_id <= pids[i] && pids[i] <= max_priv_process_id) {
```

Comment:
If you do this before launch any processes, can you just assume they're all KIPs and therefore privileged?

Reply:
You can, I guess, yeah. I kind of prefer the check but it is admittedly superfluous.